Add option for Append Processor to skip/allow empty values

limotova opened the pull request:
If one is using templates to generate values, this option can be helpful to avoid adding values that resolve into empty strings

Fixes #104813

limotova committed:
Add option for Append Processor to skip/allow empty values

If one is using templates to generate values, this option can be helpful
to avoid adding values that resolve into empty strings

Fixes #104813

joegallo reviewed (changes requested):
I'd like to have this mirror the `set` processor's `ignore_empty_value` option, rather than creating a new option name with no precedent -- it's functionally the same as that option, right (albeit with the opposite sense)?

limotova committed:
Change allow_empty_values to ignore_empty_values

romulets commented:
This feature is actually pretty useful. I'm my use-case, I'm appending a lot of fields. To keep checking if they are not null on context is pretty repetitive. Right now we have a painless script that removes all empty values. But would be awesome to have this out of the box.

limotova committed:
Update docs/changelog/105718.yaml

elasticsearchmachine commented:
Hi @limotova, I've created a changelog YAML for you.

joegallo committed:
Merge branch 'main' into add-allow-empty-values-append-processor

joegallo committed:
Handle nulls as well

The original version of this PR predated `copy_from`, so the value
wouldn't have been able to be null here. But now that `copy_from`
exists we should handle null as well as empty strings.

joegallo committed:
Rewrite setFieldValue in terms of valueNotEmpty

Since it handles nulls now, this logic is identical.

joegallo committed:
Rewrite another setFieldValue arity, too

This one is a bit trickier, so I've lampshaded the surprising part
with a comment, and also there's a newly added test on main via

joegallo committed:
Verify an invariant

This arity of setFieldValue doesn't handle the single-value
ignore_empty_value case (like in a set processor), it only handles the
multi-value ignore_empty_values case (like in an append
processor). The same thing applies to allow_duplicates, that is also
an append-only concern.

nik9000 commented:
I opened an issue: https://github.com/elastic/elasticsearch/issues/134244

joegallo committed:
This test can't handle nulls anymore

joegallo committed:
Fix a typo

joegallo committed:
Declare these once

and don't use the return value from processor.execute

joegallo committed:
Add some empty values into this test

joegallo committed:
Merge branch 'main' into add-allow-empty-values-append-processor

limotova reviewed (commented):
I can't approve my own PR haha but the changes look reasonable to me, thank you!

joegallo committed:
Merge branch 'main' into add-allow-empty-values-append-processor

joegallo committed:
Ignore empty values with getFieldValue and copy_from

joegallo committed:
Re-add the ignore_empty_values docs

joegallo committed:
It seems conventional to just label these '9.2'

joegallo committed:
Add an example

joegallo committed:
Better quoting

joegallo committed:
Add a test-only feature

joegallo committed:
Add some rest tests

joegallo committed:
Fix typos

joegallo committed:
Merge branch 'main' into add-allow-empty-values-append-processor
